tests/src/common/test_case/address_book.rs

```diff
@@ -0,0 +1,235 @@
+    AddressBook<Dispatcher, Treasury, Profit, Leaser, (), Oracle, TimeAlarms>
+{
+    pub(super) fn with_lpp(
+        self,
+        lpp_addr: Addr,
+    ) -> AddressBook<Dispatcher, Treasury, Profit, Leaser, Addr, Oracle, TimeAlarms> {
+        AddressBook {
+            dispatcher_addr: self.dispatcher_addr,
+            treasury_addr: self.treasury_addr,
+            profit_addr: self.profit_addr,
+            leaser_addr: self.leaser_addr,
+            lpp_addr,
+            oracle_addr: self.oracle_addr,
+            time_alarms_addr: self.time_alarms_addr,
+            lease_code_id: self.lease_code_id,
+        }
+    }
+}
+
+impl<Dispatcher, Treasury, Profit, Leaser, Oracle, TimeAlarms>
+    AddressBook<Dispatcher, Treasury, Profit, Leaser, Addr, Oracle, TimeAlarms>
+{
+    pub const fn lpp(&self) -> &Addr {
+        &self.lpp_addr
+    }
+}
+
+impl<Dispatcher, Treasury, Profit, Leaser, Lpp, TimeAlarms>
+    AddressBook<Dispatcher, Treasury, Profit, Leaser, Lpp, (), TimeAlarms>
+{
+    pub(super) fn with_oracle(
+        self,
+        oracle_addr: Addr,
+    ) -> AddressBook<Dispatcher, Treasury, Profit, Leaser, Lpp, Addr, TimeAlarms> {
+        AddressBook {
+            dispatcher_addr: self.dispatcher_addr,
+            treasury_addr: self.treasury_addr,
+            profit_addr: self.profit_addr,
+            leaser_addr: self.leaser_addr,
+            lpp_addr: self.lpp_addr,
+            oracle_addr,
+            time_alarms_addr: self.time_alarms_addr,
+            lease_code_id: self.lease_code_id,
+        }
+    }
+}
+
+impl<Dispatcher, Treasury, Profit, Leaser, Lpp, TimeAlarms>
+    AddressBook<Dispatcher, Treasury, Profit, Leaser, Lpp, Addr, TimeAlarms>
+{
+    pub const fn oracle(&self) -> &Addr {
+        &self.oracle_addr
+    }
+}
+
+impl<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle>
+    AddressBook<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle, ()>
+{
+    pub(super) fn with_time_alarms(
+        self,
+        time_alarms_addr: Addr,
+    ) -> AddressBook<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle, Addr> {
+        AddressBook {
+            dispatcher_addr: self.dispatcher_addr,
+            treasury_addr: self.treasury_addr,
+            profit_addr: self.profit_addr,
+            leaser_addr: self.leaser_addr,
+            lpp_addr: self.lpp_addr,
+            oracle_addr: self.oracle_addr,
+            time_alarms_addr,
+            lease_code_id: self.lease_code_id,
+        }
+    }
+}
+
+impl<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle>
+    AddressBook<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle, Addr>
+{
+    pub const fn time_alarms(&self) -> &Addr {
+        &self.time_alarms_addr
+    }
+}
+
+impl<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle, TimeAlarms>
+    AddressBook<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle, TimeAlarms>
+{
+    pub(super) fn set_lease_code_id(&mut self, lease_code_id: u64) {
+        self.lease_code_id = lease_code_id;
+    }
```

Comment:
please return a new instance than mutating self

Reply:
I did but then the problem occurred when calling `TestCase::store_new_lease_code` because it is behind a mutable reference and doesn't allow me to move it out.